llrt_core/src/modules/encoding/text_encoder.rs

```diff
@@ -21,4 +24,34 @@ impl TextEncoder {
     pub fn encode<'js>(&self, ctx: Ctx<'js>, string: String) -> Result<Value<'js>> {
         TypedArray::new(ctx, string.as_bytes()).map(|m| m.into_value())
     }
+
+    pub fn encode_into<'js>(
+        &self,
+        ctx: Ctx<'js>,
+        string: String,
+        obj: Object<'js>,
+    ) -> Result<Object<'js>> {
+        let mut read = 0;
+        let mut written = 0;
+
+        if let Some((array_buffer, source_length, source_offset)) = obj_to_array_buffer(&obj)? {
+            let raw = array_buffer
+                .as_raw()
+                .ok_or("ArrayBuffer is detached")
+                .or_throw(&ctx)?;
+
+            let bytes = unsafe {
+                std::slice::from_raw_parts_mut(raw.ptr.as_ptr().add(source_offset), source_length)
+            };
+
+            let mut enc = encoding_rs::UTF_8.new_encoder();
+            (_, _, written, _) = enc.encode_from_utf8(string.as_str(), bytes, false);
+            read = from_utf8(&bytes[..written]).unwrap().encode_utf16().count();
```

Comment:
This seems inefficient as we're encoding twice. Can we do encoding only once?

Reply:
Does "twice" mean `from_utf8()` and `encode_utf16()`?
If so, we cannot avoid it for the following reasons at the moment.

1. To correctly calculate the number of units in UTF-16, at this time we need to run `encode_utf16()`.
2. And `encode_utf16()` must be performed on `&str`.
3. In addition, `&bytes[. .written]` is `&[u8]`, we need to run `from_utf8()` to convert from `&[u8]` to `&str`.

I may be overlooking something that could be done more easily.

Reply:
Would something like this work? Arn't rust strings always utf8?
```rust
pub fn encode_into<'js>(
    &self,
    ctx: Ctx<'js>,
    string: String,
    obj: Object<'js>,
) -> Result<Object<'js>> {
    let mut read = 0;
    let mut written = 0;

    if let Some((array_buffer, source_length, source_offset)) = obj_to_array_buffer(&obj)? {
        let raw = array_buffer
            .as_raw()
            .ok_or("ArrayBuffer is detached")
            .or_throw(&ctx)?;
        let bytes = unsafe {
            std::slice::from_raw_parts_mut(raw.ptr.as_ptr().add(source_offset), source_length)
        };

    
       read = string.chars().fold(0, |acc, ch| acc + ch.len_utf16());

        let utf8_bytes = string.as_bytes();
        written = utf8_bytes.len().min(bytes.len());
        bytes[..written].copy_from_slice(&utf8_bytes[..written]);
    }

    let obj = Object::new(ctx)?;
    obj.set("read", read)?;
    obj.set("written", written)?;
    Ok(obj)
}
```

Reply:
I tried it immediately, and there are some things that worked and some things that did not work.

```rust
written = utf8_bytes.len().min(bytes.len());
```
`wrtten` must return a valid number of digits when rounded out by that number of digits as UTF-8. This means that the number could be even smaller than the two minimums.
This decision is not easy, as it requires consideration of multibyte characters and surrogate pairs.
Therefore, I had to rely on `decoding_rs::UTF_8::encode_from_utf8()` to make this determination.

```rust
read = string.chars().fold(0, |acc, ch| acc + ch.len_utf16());
```
Assuming that `written` can be calculated correctly, we were able to incorporate this implementation by cutting out the string by the number of bytes indicated by `written`.